File: src/main/java/se/kth/spork/spoon/conflict/OptimisticInsertInsertHandler.java
package se.kth.spork.spoon.conflict;

import java.util.List;
import java.util.Optional;
import se.kth.spork.spoon.wrappers.SpoonNode;

/**
 * A structural conflict handler that optimistically resolves insert/insert conflicts in which one
 * side is empty by choosing the non-empty side.
 *
 * @author Simon Larsén
 */
public class OptimisticInsertInsertHandler implements StructuralConflictHandler {
    @Override
    public Optional<List<SpoonNode>> tryResolveConflict(
            List<SpoonNode> leftNodes, List<SpoonNode> rightNodes, ConflictType type) {
        if (!(leftNodes.isEmpty() || rightNodes.isEmpty()) || type != ConflictType.INSERT_INSERT)
            return Optional.empty();
        return Optional.of(leftNodes.isEmpty() ? rightNodes : leftNodes);
    }
}

File: src/main/kotlin/se/kth/spork/spoon/conflict/OptimisticInsertInsertHandler.kt
package se.kth.spork.spoon.conflict

import se.kth.spork.spoon.wrappers.SpoonNode

/**
 * A structural conflict handler that optimistically resolves insert/insert conflicts in which one
 * side is empty by choosing the non-empty side.
 *
 * @author Simon Larsén
 */
class OptimisticInsertInsertHandler : StructuralConflictHandler {
    override fun tryResolveConflict(
        leftNodes: List<SpoonNode>,
        rightNodes: List<SpoonNode>,
        type: ConflictType,
    ): List<SpoonNode>? {
        return if (leftNodes.isNotEmpty() && rightNodes.isNotEmpty() || type != ConflictType.INSERT_INSERT) {
            null
        } else if (leftNodes.isNotEmpty()) {
            leftNodes
        } else {
            rightNodes
        }
    }
}
